CDAP-7721 TMS TTL Cleanup in HBase and LevelDB tables

gokulavasan opened the pull request:
JIRA : https://issues.cask.co/browse/CDAP-7721
Build : http://builds.cask.co/browse/CDAP-RUT301

Summary:

* Adds cdap-tms-tests module since hbase-compat modules now depends on cdap-tms and thus for tests, we need to have a separate module so that there is no circular dependency (similar to cdap-data-fabric/cdap-data-fabric-tests modules)

* Adding two coprocessors for Transaction Messaging Service Data cleanup in HBase tables.
One for MessageTable and another for PayloadTable.
The logic in the coprocessor queries the metadata table for the TTL value and then checks if the TTL has elapsed, in which case, it skips the cell. Otherwise, it retains the cell.

* LevelDB implementation (scheduled thread that invokes data pruning for each topic on both message table and payload table). Tests are currently timing sensitive and thus might be flaky. Any thoughts how to make it more stable are welcome.

Note: This PR is same as the old one here https://github.com/caskdata/cdap/pull/7179 but with the comments on the old PR addressed in this PR. The old PR was closed since it was opened against develop.

gokulavasan committed:
CDAP-7409 TMS HBase tables data cleanup coprocessors

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/conf/Constants.java:
```diff
@@ -1106,6 +1107,11 @@
+    public static final String HBASE_METADATA_TABLE_NAME = "cdap.messaging.metadata.table.name";
+    public static final String HBASE_METADATA_TABLE_NAMESPACE = "cdap.messaging.metadata.hbase.namespace";
+
+    public static final String COPROCESSOR_DIR = "messaging.coprocessor.dir";
```
Can you move this up together with other cConf property keys? Also add a comment for the `HBASE_METADATA_TABLE_*` constants, as they are not cConf keys, but rather HBase table attributes (hence need the `cdap.` prefix, I believe).

chtyim commented on cdap-common/src/main/resources/cdap-default.xml:
```diff
@@ -1137,6 +1137,12 @@
   <property>
+    <name>messaging.local.data.ttl.cleanup.frequency.secs</name>
+    <value>3600</value>
+    <description>Scheduling frequency of TTL cleanup thread in seconds (only used in Standalone CDAP)</description>
```
I think @awholegunch requires it formatted as:

```xml
<description>
  The description goes here.
</description>

chtyim commented on cdap-hbase-compat-0.98/src/main/java/co/cask/cdap/data2/transaction/messaging/coprocessor/hbase98/MessageTableRegionObserver.java:
```diff
@@ -0,0 +1,166 @@
+                            earliestPutTs);
+  }
+
+  private final class TTLFilter extends FilterBase {
```
Can this be static class?

chtyim commented on cdap-hbase-compat-0.98/src/main/java/co/cask/cdap/data2/transaction/messaging/coprocessor/hbase98/MessageTableRegionObserver.java:
```diff
@@ -0,0 +1,166 @@
+  private static final byte[] COL_FAMILY = MessagingUtils.Constants.COLUMN_FAMILY;
+  private static final byte[] COL = MessagingUtils.Constants.METADATA_COLUMN;
+
+  private static final int PREFIX_LENGTH = MessagingUtils.Constants.TABLE_PREFIX_NUM_BYTES;
```
I remember I commented on this before, but on a 2nd through, it's probably better to be in the table attribute that get set when the table was created.

chtyim commented on cdap-hbase-compat-0.98/src/main/java/co/cask/cdap/data2/transaction/messaging/coprocessor/hbase98/MessageTableRegionObserver.java:
```diff
@@ -0,0 +1,166 @@
+      int rowKeyOffset = cell.getRowOffset() + PREFIX_LENGTH;
+      int sizeOfRowKey = cell.getRowLength() - PREFIX_LENGTH;
+      long publishTimestamp = MessagingUtils.getPublishTimestamp(cell.getRowArray(), rowKeyOffset, sizeOfRowKey);
+      byte[] topicIdBytes = Arrays.copyOfRange(cell.getRowArray(), rowKeyOffset,
```
Try not the copy the topic id bytes for each cell. You can do the comparison using `public static int compareTo(byte[] buffer1, int offset1, int length1, byte[] buffer2, int offset2, int length2)` in the HBase `Bytes` class.

chtyim commented on cdap-hbase-compat-0.98/src/main/java/co/cask/cdap/data2/transaction/messaging/coprocessor/hbase98/MessageTableRegionObserver.java:
```diff
@@ -0,0 +1,166 @@
+                                               rowKeyOffset + MessagingUtils.getTopicLengthMessageEntry(sizeOfRowKey));
+      try {
+        if (prevTopicIdBytes == null || (!Arrays.equals(prevTopicIdBytes, topicIdBytes)) || currentTTL == null) {
+          currentTTL = topicCache.get(topicIdBytes);
```
`byte[]` doesn't support `equals` and `hashCode` as required by `LoadingCache` (which is a `map`), meaning this call will always try to load the topic info. Better use a `ByteBuffer` to wrap row array with offset and len as `ByteBuffer` supports `equals` and `hashCode` correctly.

chtyim commented on cdap-hbase-compat-0.98/src/main/java/co/cask/cdap/data2/transaction/messaging/coprocessor/hbase98/MessageTableRegionObserver.java:
```diff
@@ -0,0 +1,166 @@
+          return ReturnCode.SKIP;
+        }
+      } catch (ExecutionException ex) {
+        LOG.debug("Region {}" + env.getRegion().getRegionNameAsString() + ", exception while" +
```
I don't think the common logger supports `{}` and you are already using string concatenation.

chtyim commented on cdap-hbase-compat-0.98/src/main/java/co/cask/cdap/data2/transaction/messaging/coprocessor/hbase98/PayloadTableRegionObserver.java:
```diff
@@ -0,0 +1,166 @@
+ *
+ * TODO: Add logic to remove data corresponding to invalidated transactions
+ */
+public class PayloadTableRegionObserver extends BaseRegionObserver {
```
Similar comments on the `MessageTableRegionObserver` applies here as well.

gokulavasan replied:
Implemented the changes suggested above in PayloadTableRegionObserver as well.

gokulavasan commented:
@chtyim Implemented all the cleanup logic required in HBase coprocessors and LevelDB. Added unit tests for all the cases as well. Please take a look.

gokulavasan commented:
@chtyim I have implemented all the clean up logic as described here https://wiki.cask.co/display/CE/Messaging#Messaging-DataCleanupInTables. I have added unit tests for all the cases - HBase and LevelDB. Once you review, I will port the coprocessors to other compat modules, run the build and then squash and merge.

chtyim commented on cdap-hbase-compat-0.98/src/main/java/co/cask/cdap/data2/transaction/messaging/coprocessor/hbase98/MessageTableRegionObserver.java:
```diff
@@ -0,0 +1,228 @@
+    if (env instanceof RegionCoprocessorEnvironment) {
+      HTableDescriptor tableDesc = ((RegionCoprocessorEnvironment) env).getRegion().getTableDesc();
+      String metadataTableNamespace = tableDesc.getValue(Constants.MessagingSystem.HBASE_METADATA_TABLE_NAMESPACE);
+      String metadataTableName = tableDesc.getValue(Constants.MessagingSystem.HBASE_METADATA_TABLE_NAME);
```
Maybe better just read it off from cConfReader.

gokulavasan replied:
I have modified this so that we read the metadataTableName from cConf. However, it is not straightforward to remove metadataTableNamespace attribute. This is because we use tableUtil to convert the cdap namespace to hbase namespace and then we write to the data table as an attribute. We can also get HBaseTableUtil in the coprocessor and do the conversion (and it will work since for system namespace we don't look up NamespaceQueryAdmin) but we can avoid it if we use the attribute.

chtyim commented on cdap-hbase-compat-0.98/src/main/java/co/cask/cdap/data2/transaction/messaging/coprocessor/hbase98/MessageTableRegionObserver.java:
```diff
@@ -0,0 +1,228 @@
+
+          @Override
+          public Map<String, String> load(ByteBuffer topicBytes) throws Exception {
+            Get get = new Get(topicBytes.array());
```
This is not entirely correct. Use hbase `Bytes.toBytes(ByteBuffer buf)` instead.

gokulavasan replied:
I thought BytesBuffer#array just returns the backing byte array. And when the ByteBuffer is created the entire prevTopicIdBytes is used as the backing byte array.

chtyim replied:
This part is dangerous because the part that constructs this `ByteBuffer` is separated from here and if there is any change in the construction, this will break. You can use `array()` safe if and only if `buffer.remaining() == buffer.array().length`

chtyim commented on cdap-hbase-compat-0.98/src/main/java/co/cask/cdap/data2/transaction/messaging/coprocessor/hbase98/MessageTableRegionObserver.java:
```diff
@@ -0,0 +1,228 @@
+
+  private static final Log LOG = LogFactory.getLog(MessageTableRegionObserver.class);
+  private static final Gson GSON = new Gson();
+  private static final Type MAP_TYPE = new TypeToken<SortedMap<String, String>>() { }.getType();
```
This can be just `Map` instead of `SortedMap` because we are just reading.

chtyim commented on cdap-hbase-compat-0.98/src/main/java/co/cask/cdap/data2/transaction/messaging/coprocessor/hbase98/MessageTableRegionObserver.java:
```diff
@@ -0,0 +1,228 @@
+            byte[] properties = result.getValue(COL_FAMILY, COL);
+            Map<String, String> propertyMap = GSON.fromJson(Bytes.toString(properties), MAP_TYPE);
+            String ttl = propertyMap.get(MessagingUtils.Constants.TTL_KEY);
+            Long ttlInMs = TimeUnit.SECONDS.toMillis(Long.parseLong(ttl));
```
should be `long`.

chtyim commented on cdap-hbase-compat-0.98/src/main/java/co/cask/cdap/data2/transaction/messaging/coprocessor/hbase98/MessageTableRegionObserver.java:
```diff
@@ -0,0 +1,228 @@
+  private static final byte[] COL_FAMILY = MessagingUtils.Constants.COLUMN_FAMILY;
+  private static final byte[] COL = MessagingUtils.Constants.METADATA_COLUMN;
+
+  private static int prefixLength;
```
why these are static?

chtyim commented on cdap-hbase-compat-0.98/src/main/java/co/cask/cdap/data2/transaction/messaging/coprocessor/hbase98/MessageTableRegionObserver.java:
```diff
@@ -0,0 +1,228 @@
+    MessageDataFilter(RegionCoprocessorEnvironment env, long timestamp) {
+      this.env = env;
+      this.timestamp = timestamp;
+      this.state = txStateCache.getLatestState();
```
Good to log some information about the state. E.g. the timestamp of the snapshot and the visibility bound. It helps debug when need to answer the question of why some entries are not just removed.

chtyim commented on cdap-hbase-compat-0.98/src/main/java/co/cask/cdap/data2/transaction/messaging/coprocessor/hbase98/MessageTableRegionObserver.java:
```diff
@@ -0,0 +1,228 @@
+
+      try {
+        if (prevTopicIdBytes == null || currentTTL == null || currentGen == null ||
+          (!org.apache.hadoop.hbase.util.Bytes.equals(prevTopicIdBytes, 0, prevTopicIdBytes.length,
```
Why need full name? For coprocessor, we should always use the hbase `Bytes`.

chtyim commented on cdap-hbase-compat-0.98/src/main/java/co/cask/cdap/data2/transaction/messaging/coprocessor/hbase98/MessageTableRegionObserver.java:
```diff
@@ -0,0 +1,228 @@
+        }
+
+        // TTL expiration cleanup only if the generation of metadata and row key are the same
+        if ((generationId == Math.abs(currentGen)) && ((timestamp - publishTimestamp) > currentTTL)) {
```
No need to have `Math.abs` here. This is because if `currentGen < 0` and `dataGen == Math.abs(currentGen)`, it is already skipped by the previous `isOlderGeneration` condition.

chtyim commented on cdap-hbase-compat-0.98/src/main/java/co/cask/cdap/data2/transaction/messaging/coprocessor/hbase98/MessageTableRegionObserver.java:
```diff
@@ -0,0 +1,228 @@
+          return ReturnCode.SKIP;
+        }
+      } catch (ExecutionException ex) {
+        LOG.debug("Region " + env.getRegion().getRegionNameAsString() + ", exception while" +
```
You might want to log a `INFO` without the stack trance with but `ex.getMessage()`, followed by logging the stacktrace using debug.

chtyim commented on cdap-hbase-compat-0.98/src/main/java/co/cask/cdap/data2/transaction/messaging/coprocessor/hbase98/MessageTableRegionObserver.java:
```diff
@@ -0,0 +1,228 @@
+        }
+      } catch (ExecutionException ex) {
+        LOG.debug("Region " + env.getRegion().getRegionNameAsString() + ", exception while" +
+                    "trying to fetch properties of topicId {}" + MessagingUtils.toTopicId(prevTopicIdBytes), ex);
```
I don't know `{}` is supported in common logging framework.

chtyim commented on cdap-hbase-compat-0.98/src/main/java/co/cask/cdap/data2/transaction/messaging/coprocessor/hbase98/MessageTableRegionObserver.java:
```diff
@@ -0,0 +1,228 @@
+    public Cell transformCell(Cell cell) throws IOException {
+      // Rollback expired/invalidated transactions
+      byte[] txCol = MessagingUtils.Constants.TX_COL;
+      if (org.apache.hadoop.hbase.util.Bytes.equals(cell.getQualifierArray(), cell.getQualifierOffset(),
```
Same here, no need to use full name.

chtyim commented on cdap-hbase-compat-0.98/src/main/java/co/cask/cdap/data2/transaction/messaging/coprocessor/hbase98/MessageTableRegionObserver.java:
```diff
@@ -0,0 +1,228 @@
+                                                    cell.getQualifierLength(), txCol, 0, txCol.length)) {
+        // This is a TX COL
+        long txWritePtr = Bytes.toLong(cell.getValueArray(), cell.getValueOffset());
+        if (txWritePtr > 0 && state != null && state.getInvalid().contains(txWritePtr)) {
```
The invalid list is always sorted, so instead of using `contains`, use binary search.

gokulavasan replied:
state.getInvalid() returns a Collection<long> and thus it is not easy to implement binary search. One way would be to 'getArray' from Collection but that would involve allocation memory to construct the byte array.

chtyim replied:
It is actually a list. So you can do a small optimization by first check if it is instance of `List`. If it is, then cast and use `Collections.binarySearch`, otherwise fallback to `contains` method.

chtyim commented on cdap-tms/src/main/java/co/cask/cdap/messaging/store/leveldb/LevelDBTableFactory.java:
```diff
@@ -99,4 +114,26 @@ private File ensureDirExists(File dir) throws IOException {
+
+      long timeStamp = System.currentTimeMillis();
+      try {
+        List<TopicId> topicIdList = metadataTable.listTopics();
```
Instead of just list existing topics, you can add a method to `LevelDBMetadataTable` to iterates over all topics. Something like:

`CloseableIterator<TopicMetadata> scanTopics(boolean includeDeleted) throws IOException`

chtyim commented on cdap-hbase-compat-0.98/src/main/java/co/cask/cdap/data2/transaction/messaging/coprocessor/hbase98/MessageTableRegionObserver.java:
```diff
@@ -0,0 +1,239 @@
+
+          @Override
+          public Map<String, String> load(ByteBuffer topicBytes) throws Exception {
+            Get get = new Get(Bytes.toBytes(topicBytes));
```
Well, you can do a small optimization like `topicBytes.array().length == topicBytes.remaining() ? topicBytes.array() : Bytes.toBytes(topicBytes)`

gokulavasan commented:
@chtyim Addressed comments in the last commit. Please take a look again when you get a chance. Thanks!

chtyim commented on cdap-hbase-compat-0.98/src/main/java/co/cask/cdap/data2/transaction/messaging/coprocessor/hbase98/MessageTableRegionObserver.java:
```diff
@@ -96,16 +97,20 @@ public void start(CoprocessorEnvironment env) throws IOException {
-      String sysConfigTablePrefix = nameConverter.getSysConfigTablePrefix(hbaseNamespacePrefix);
-      cConfReader = new CConfigurationReader(env.getConfiguration(), sysConfigTablePrefix);
-      final Long metadataCacheExpiry = Long.parseLong(cConfReader.read().get(
+      final Long metadataCacheExpiry = Long.parseLong(cConf.get(
```
`cConf` has a `getLong` method that will do the parsing. Also, the type should be `long`, not `Long`

chtyim commented on cdap-tms/src/main/java/co/cask/cdap/messaging/store/leveldb/LevelDBMetadataTable.java:
```diff
@@ -158,32 +159,63 @@ public void deleteTopic(TopicId topicId) throws TopicNotFoundException, IOExcept
-  @Override
-  public void close() throws IOException {
-    // no op
+  public CloseableIterator<TopicMetadata> scanTopics() throws IOException {
```
Add a javadoc to state clear that it returns all topics, including deleted one

chtyim commented:
Only two comments left, the rest LGTM

chtyim commented:
LGTM
